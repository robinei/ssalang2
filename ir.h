#pragma once

#include "defs.h"

typedef i16 IrOperand;

typedef IrOperand IrInstrRef; // signed, so that pure instructions can have negative index (that way we don't commit to any sequencing of them early)
typedef IrOperand IrBlockRef;

typedef enum IrType {
  TY_VOID,
  TY_BOOL,
  TY_I32,
} IrType;


// annotated with types for (arg0[, arg1[, arg2]])
typedef enum IrInstrTag {
  IR_NOP,
  IR_IDENTITY, // (IrInstrRef replace_with)
  
  IR_LABEL, // (IrBlockRef block)

  // terminals
  IR_JUMP, // (IrBlockRef target)
  IR_BRANCH, // (IrInstrRef cond, IrBlockRef true_target, IrBlockRef false_target)
  IR_RET, // (IrInstrRef value)

  IR_UPSILON, // (IrInstrRef value, IrInstrRef phi)

  // all after this point must be pure
  IR_CONST,
  IR_PHI, // (IrOperand unique_phi_id)
  IR_ARG, // (IrOperand argument_index)
  IR_ADD, // (IrInstrRef lhs, IrInstrRef rhs)
  IR_EQ, // (IrInstrRef lhs, IrInstrRef rhs)
} IrInstrTag;

#define IR_IS_PURE_INSTR(x) ((x) >= IR_CONST)
#define IR_IS_TERMINAL_INSTR(x) ((x) >= IR_JUMP && (x) <= IR_RET)

typedef union IrInstr {
  struct {
    IrInstrTag tag : 8;
    IrType type : 8;
    IrOperand arg0 : 16;
    union {
      struct {
        IrOperand arg1, arg2;
      };
      bool bool_const;
      i32 i32_const;
      u32 u32_const;
      f32 f32_const;
    };
  };
  u64 u64_repr;
} IrInstr;
