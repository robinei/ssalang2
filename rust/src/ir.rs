use std::num::NonZeroI16;

pub type Operand = i16;

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub struct InstrRef(pub NonZeroI16);

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub struct BlockRef(pub NonZeroI16);

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub struct PhiRef(pub NonZeroI16);

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub struct VarRef(pub NonZeroI16);

impl InstrRef {
    pub fn new(value: Operand) -> Option<Self> {
        NonZeroI16::new(value).map(Self)
    }

    pub fn get(self) -> Operand {
        self.0.get()
    }
}

impl BlockRef {
    pub fn new(value: Operand) -> Option<Self> {
        NonZeroI16::new(value).map(Self)
    }

    pub fn get(self) -> Operand {
        self.0.get()
    }
}

impl PhiRef {
    pub fn new(value: Operand) -> Option<Self> {
        NonZeroI16::new(value).map(Self)
    }

    pub fn get(self) -> Operand {
        self.0.get()
    }
}

impl VarRef {
    pub fn new(value: Operand) -> Option<Self> {
        NonZeroI16::new(value).map(Self)
    }

    pub fn get(self) -> Operand {
        self.0.get()
    }
}

impl From<InstrRef> for Operand {
    fn from(value: InstrRef) -> Self {
        value.get()
    }
}

impl From<BlockRef> for Operand {
    fn from(value: BlockRef) -> Self {
        value.get()
    }
}

impl From<PhiRef> for Operand {
    fn from(value: PhiRef) -> Self {
        value.get()
    }
}

impl From<VarRef> for Operand {
    fn from(value: VarRef) -> Self {
        value.get()
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
#[repr(u8)]
pub enum Type {
    Void = 0,
    Bool = 1,
    I32 = 2,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub struct Meta(u8);

impl Meta {
    pub fn new(ty: Type) -> Self {
        Self(ty as u8)
    }
    
    pub fn get_type(self) -> Type {
        match self.0 & 0x7F {
            0 => Type::Void,
            1 => Type::Bool,
            2 => Type::I32,
            _ => unreachable!(),
        }
    }
    
    pub fn is_marked(self) -> bool {
        (self.0 & 0x80) != 0
    }
    
    pub fn set_marked(&mut self, marked: bool) {
        if marked {
            self.0 |= 0x80;
        } else {
            self.0 &= 0x7F;
        }
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
#[repr(u8)]
pub enum Instr {
    Nop(Meta),
    Identity(Meta, InstrRef),
    Print(Meta, InstrRef),

    // CFG instructions
    Label(Meta, BlockRef),
    Jump(Meta, BlockRef),
    Branch(Meta, InstrRef, BlockRef, BlockRef),
    Ret(Meta, InstrRef),

    // SSA annotations
    Upsilon(Meta, PhiRef, InstrRef),
    Phi(Meta, PhiRef),

    // Pure instructions
    ConstBool(Meta, bool),
    ConstI32(Meta, i32),
    Arg(Meta, i32),
    Add(Meta, InstrRef, InstrRef),
    Eq(Meta, InstrRef, InstrRef),
    Neq(Meta, InstrRef, InstrRef),
}

impl Instr {
    pub fn nop() -> Self {
        Self::Nop(Meta::new(Type::Void))
    }

    pub fn const_bool(value: bool) -> Self {
        Self::ConstBool(Meta::new(Type::Bool), value)
    }

    pub fn const_i32(value: i32) -> Self {
        Self::ConstI32(Meta::new(Type::I32), value)
    }
    
    pub fn is_pure(&self) -> bool {
        matches!(self,
            Self::ConstBool(..) | 
            Self::ConstI32(..) | 
            Self::Arg(..) | 
            Self::Add(..) | 
            Self::Eq(..) | 
            Self::Neq(..)
        )
    }

    pub fn is_terminal(&self) -> bool {
        matches!(self, Self::Jump(..) | Self::Branch(..) | Self::Ret(..))
    }

    pub fn get_meta(&self) -> Meta {
        match self {
            Self::Nop(meta) => *meta,
            Self::Identity(meta, ..) => *meta,
            Self::Print(meta, ..) => *meta,
            Self::Label(meta, ..) => *meta,
            Self::Jump(meta, ..) => *meta,
            Self::Branch(meta, ..) => *meta,
            Self::Ret(meta, ..) => *meta,
            Self::Upsilon(meta, ..) => *meta,
            Self::Phi(meta, ..) => *meta,
            Self::ConstBool(meta, ..) => *meta,
            Self::ConstI32(meta, ..) => *meta,
            Self::Arg(meta, ..) => *meta,
            Self::Add(meta, ..) => *meta,
            Self::Eq(meta, ..) => *meta,
            Self::Neq(meta, ..) => *meta,
        }
    }

    pub fn get_type(&self) -> Type {
        self.get_meta().get_type()
    }

    pub fn is_marked(&self) -> bool {
        self.get_meta().is_marked()
    }
}

impl Default for Instr {
    fn default() -> Self {
        Self::nop()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_instr_size() {
        // With packed Meta (1 byte), the largest variant (Branch) needs:
        // tag(1) + meta(1) + 3*Operand(6) = 8 bytes
        assert_eq!(std::mem::size_of::<Instr>(), 8);
        assert_eq!(std::mem::size_of::<Meta>(), 1);
        assert_eq!(std::mem::size_of::<Operand>(), 2);
    }

    #[test]
    fn test_meta() {
        let meta1 = Meta::new(Type::I32);
        assert_eq!(meta1.get_type(), Type::I32);
        assert!(!meta1.is_marked());
        
        let mut meta2 = Meta::new(Type::Bool);
        meta2.set_marked(true);
        assert_eq!(meta2.get_type(), Type::Bool);
        assert!(meta2.is_marked());
        
        let mut meta3 = Meta::new(Type::Void);
        assert!(!meta3.is_marked());
        meta3.set_marked(true);
        assert!(meta3.is_marked());
        meta3.set_marked(false);
        assert!(!meta3.is_marked());
    }

    #[test]
    fn test_bool_const() {
        let instr_true = Instr::ConstBool(
            Meta::new(Type::Bool),
            true,
        );
        
        let instr_false = Instr::ConstBool(
            Meta::new(Type::Bool),
            false,
        );
        
        if let Instr::ConstBool(_, value) = instr_true {
            assert_eq!(value, true);
        } else {
            panic!("Expected ConstBool variant");
        }
        
        if let Instr::ConstBool(_, value) = instr_false {
            assert_eq!(value, false);
        } else {
            panic!("Expected ConstBool variant");
        }
    }

    #[test]
    fn test_i32_const() {
        let instr = Instr::ConstI32(
            Meta::new(Type::I32),
            0x12345678,
        );
        
        if let Instr::ConstI32(_, value) = instr {
            assert_eq!(value, 0x12345678);
        } else {
            panic!("Expected ConstI32 variant");
        }
        
        let instr_neg = Instr::ConstI32(
            Meta::new(Type::I32),
            -1,
        );
        
        if let Instr::ConstI32(_, value) = instr_neg {
            assert_eq!(value, -1);
        } else {
            panic!("Expected ConstI32 variant");
        }
    }

    #[test]
    fn test_instr_methods() {
        let mut meta = Meta::new(Type::Void);
        meta.set_marked(true);
        let nop = Instr::Nop(meta);
        
        assert_eq!(nop.get_type(), Type::Void);
        assert!(nop.is_marked());
        assert!(!nop.is_pure());
        assert!(!nop.is_terminal());
        
        // Test get_meta method
        let meta = nop.get_meta();
        assert_eq!(meta.get_type(), Type::Void);
        assert!(meta.is_marked());
        
        let add = Instr::Add(
            Meta::new(Type::I32),
            InstrRef::new(1).unwrap(),
            InstrRef::new(2).unwrap(),
        );
        
        assert!(add.is_pure());
        assert!(!add.is_terminal());
        assert!(!add.is_marked());
        assert_eq!(add.get_meta().get_type(), Type::I32);
        
        let jump = Instr::Jump(
            Meta::new(Type::Void),
            BlockRef::new(5).unwrap(),
        );
        
        assert!(!jump.is_pure());
        assert!(jump.is_terminal());
    }

    #[test]
    fn test_niche_optimization() {
        // Test that Option<Ref> has the same size as Ref due to niche optimization
        assert_eq!(std::mem::size_of::<InstrRef>(), std::mem::size_of::<Option<InstrRef>>());
        assert_eq!(std::mem::size_of::<BlockRef>(), std::mem::size_of::<Option<BlockRef>>());
        assert_eq!(std::mem::size_of::<PhiRef>(), std::mem::size_of::<Option<PhiRef>>());
        assert_eq!(std::mem::size_of::<VarRef>(), std::mem::size_of::<Option<VarRef>>());
        
        // All should be exactly 16 bits (2 bytes)
        assert_eq!(std::mem::size_of::<InstrRef>(), 2);
        assert_eq!(std::mem::size_of::<Option<InstrRef>>(), 2);
        assert_eq!(std::mem::size_of::<BlockRef>(), 2);
        assert_eq!(std::mem::size_of::<Option<BlockRef>>(), 2);
        assert_eq!(std::mem::size_of::<PhiRef>(), 2);
        assert_eq!(std::mem::size_of::<Option<PhiRef>>(), 2);
        assert_eq!(std::mem::size_of::<VarRef>(), 2);
        assert_eq!(std::mem::size_of::<Option<VarRef>>(), 2);
    }

    #[test]
    fn test_ref_types() {
        let instr_ref = InstrRef::new(42).unwrap();
        assert_eq!(Operand::from(instr_ref), 42);
        assert_eq!(instr_ref.get(), 42);
        
        let block_ref = BlockRef::new(-1).unwrap();
        assert_eq!(Operand::from(block_ref), -1);
        assert_eq!(block_ref.get(), -1);
        
        let phi_ref = PhiRef::new(100).unwrap();
        assert_eq!(Operand::from(phi_ref), 100);
        assert_eq!(phi_ref.get(), 100);
        
        // Test equality and hashing work
        assert_eq!(InstrRef::new(42).unwrap(), InstrRef::new(42).unwrap());
        assert_ne!(InstrRef::new(42).unwrap(), InstrRef::new(43).unwrap());
    }

    #[test]
    fn test_nop_constructor() {
        let nop = Instr::nop();
        assert_eq!(nop.get_type(), Type::Void);
        assert!(!nop.is_pure());
        assert!(!nop.is_terminal());
        
        // Test that Default trait uses nop constructor
        let default_instr = Instr::default();
        assert_eq!(default_instr.get_type(), Type::Void);
        
        // Verify they're the same
        assert_eq!(nop, default_instr);
    }
}
