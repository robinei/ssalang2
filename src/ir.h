#pragma once

#include "defs.h"

typedef i16 IrOperand;

typedef IrOperand IrInstrRef; // signed, so that pure instructions can have negative index (that way we don't commit to any sequencing of them early)
typedef IrOperand IrBlockRef;
typedef IrOperand IrPhiRef;
typedef IrOperand IrVarRef;

typedef enum IrType {
  TY_VOID,
  TY_BOOL,
  TY_I32,
} IrType;

#define IR_FLAG_MARK 1

// annotated with types for (arg0[, arg1[, arg2]])
typedef enum IrInstrTag {
  // IR_NOP is a no-op instruction, which will be removed later during fixup. replace an instruction with a NOP to remove it from the IR.
  IR_NOP,

  // IR_IDENTITY is used to replace an instruction with another instruction.
  // when fixing up the code later, each use of the original (IR_IDENTITY) instruction is replaced with a use of the replacement instruction.
  IR_IDENTITY, // arguments: (IrInstrRef replace_with)

  IR_PRINT, // arguments: (IrInstrRef value)
  
  // IR_LABEL annotates the start of a block
  IR_LABEL, // arguments: (IrBlockRef block)

  // terminals
  IR_JUMP, // arguments: (IrBlockRef target)
  IR_BRANCH, // arguments: (IrInstrRef cond, IrBlockRef true_target, IrBlockRef false_target)
  IR_RET, // arguments: (IrInstrRef value)

  // IR_UPSILON instructions are inserted at source blocks, and indicate which value to use for an IR_PHI instruction at the target block
  // this is slightly different to how phi arguments are usually encoded
  IR_UPSILON, // arguments: (IrPhiRef phi, IrInstrRef value)

  // IR_PHI instructions are used to merge values from different blocks
  IR_PHI, // arguments: (IrPhiRef phi)

  // all after this point must be pure instructions

  // constants which have their value stored in the instruction itself
  IR_CONST,
  
  // IR_ARG instructions represent function arguments
  IR_ARG, // arguments: (IrOperand argument_index)

  // binary operations
  IR_ADD, // arguments: (IrInstrRef lhs, IrInstrRef rhs)
  IR_EQ, // arguments: (IrInstrRef lhs, IrInstrRef rhs)
  IR_NEQ, // arguments: (IrInstrRef lhs, IrInstrRef rhs)

  IR_NUM_TAGS,
} IrInstrTag;

#define IR_IS_PURE_INSTR(x) ((x) >= IR_CONST)
#define IR_IS_TERMINAL_INSTR(x) ((x) >= IR_JUMP && (x) <= IR_RET)

typedef union IrInstr {
  struct {
    IrInstrTag tag : 8;
    IrType type : 4;
    u32 flags : 4;
    IrOperand arg0 : 16;
    union {
      struct {
        IrOperand arg1, arg2;
      };
      bool bool_const;
      i32 i32_const;
      u32 u32_const;
      f32 f32_const;
    };
  };
  u64 u64_repr;
} IrInstr;
