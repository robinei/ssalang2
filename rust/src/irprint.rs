use crate::code::Code;
use crate::ir::{Instr, InstrRef, BlockRef, PhiRef, Type, Meta};
use std::fmt::Write;

/// Printer for IR code that produces human-readable assembly-like output
pub struct IrPrinter<'a> {
    code: &'a Code,
    use_colors: bool,
    buffer: String,
}

#[derive(Clone, Copy)]
enum Color {
    Reset,
    Instruction,  // Blue
    Reference,    // Green  
    Type,         // Yellow
    Constant,     // Cyan
    Comment,      // Gray
}

impl Color {
    fn as_ansi(self) -> &'static str {
        match self {
            Color::Reset => "\x1b[0m",
            Color::Instruction => "\x1b[34m",  // Blue
            Color::Reference => "\x1b[32m",    // Green
            Color::Type => "\x1b[33m",         // Yellow
            Color::Constant => "\x1b[36m",     // Cyan
            Color::Comment => "\x1b[90m",      // Gray
        }
    }
}

impl<'a> IrPrinter<'a> {
    /// Create a new IR printer without colors
    pub fn new(code: &'a Code) -> Self {
        Self {
            code,
            use_colors: false,
            buffer: String::with_capacity(1024),
        }
    }

    /// Create a new IR printer with ANSI colors
    pub fn new_colored(code: &'a Code) -> Self {
        Self {
            code,
            use_colors: true,
            buffer: String::with_capacity(1024),
        }
    }

    /// Print the IR code and return the formatted string
    pub fn print(mut self) -> String {
        self.buffer.clear();
        
        self.print_constants();
        self.print_scheduled_code();
        
        self.buffer
    }

    fn print_constants(&mut self) {
        let mut has_constants = false;
        
        // Check if we have any negative indices (constants)
        for (instr_ref, _) in self.code.iter_with_refs() {
            if instr_ref.get() < 0 {
                has_constants = true;
                break;
            }
        }
        
        if !has_constants {
            return;
        }
        
        self.write_colored(Color::Comment, "; === CONSTANTS & UNSCHEDULED ===\n");
        
        for (instr_ref, instr) in self.code.iter_with_refs() {
            if instr_ref.get() < 0 {
                self.print_constant_instruction(instr_ref, instr);
            }
        }
        
        self.buffer.push('\n');
    }

    fn print_scheduled_code(&mut self) {
        let mut has_scheduled = false;
        
        // Check if we have any positive indices (scheduled instructions)
        for (instr_ref, _) in self.code.iter_with_refs() {
            if instr_ref.get() > 0 {
                has_scheduled = true;
                break;
            }
        }
        
        if !has_scheduled {
            return;
        }
        
        self.write_colored(Color::Comment, "; === FUNCTION CODE ===\n");
        
        for (instr_ref, instr) in self.code.iter_with_refs() {
            if instr_ref.get() > 0 {
                match instr {
                    Instr::Label(_, block_ref) => {
                        // Print block header comment, don't print the label instruction
                        let block_name = self.format_block_ref(*block_ref);
                        write!(&mut self.buffer, "; Block {}:\n", block_name).unwrap();
                    }
                    _ => {
                        self.print_scheduled_instruction(instr_ref, instr);
                    }
                }
            }
        }
    }

    fn print_constant_instruction(&mut self, instr_ref: InstrRef, instr: &Instr) {
        // Format: "c1:    const_i32  42        ; i32"
        let const_label = self.format_const_ref(instr_ref);
        self.write_colored(Color::Reference, &format!("{}:", const_label));
        self.buffer.push_str("    ");
        
        self.print_instruction_body(instr);
        self.print_type_comment(instr.get_meta());
        self.buffer.push('\n');
    }

    fn print_scheduled_instruction(&mut self, instr_ref: InstrRef, instr: &Instr) {
        // Only print instruction label for non-void instructions
        if instr.get_type() != Type::Void {
            let instr_label = self.format_instr_ref(instr_ref);
            self.write_colored(Color::Reference, &format!("{}:", instr_label));
            self.buffer.push_str("    ");
        } else {
            // Extra spaces to align with labeled instructions
            self.buffer.push_str("       ");
        }
        
        self.print_instruction_body(instr);
        
        // Only print type comment for non-void
        if instr.get_type() != Type::Void {
            self.print_type_comment(instr.get_meta());
        }
        
        self.buffer.push('\n');
    }

    fn print_instruction_body(&mut self, instr: &Instr) {
        match instr {
            Instr::Nop(_) => {
                self.write_colored(Color::Instruction, "nop");
            }
            Instr::Identity(_, operand) => {
                self.write_colored(Color::Instruction, "identity");
                self.buffer.push_str("  ");
                self.write_colored(Color::Reference, &self.format_instr_ref(*operand));
            }
            Instr::Print(_, operand) => {
                self.write_colored(Color::Instruction, "print");
                self.buffer.push_str("      ");
                self.write_colored(Color::Reference, &self.format_instr_ref(*operand));
            }
            Instr::Label(_, block_ref) => {
                self.write_colored(Color::Instruction, "label");
                self.buffer.push_str("      ");
                self.write_colored(Color::Reference, &self.format_block_ref(*block_ref));
            }
            Instr::Jump(_, block_ref) => {
                self.write_colored(Color::Instruction, "jump");
                self.buffer.push_str("       ");
                self.write_colored(Color::Reference, &self.format_block_ref(*block_ref));
            }
            Instr::Branch(_, cond, true_block, false_block) => {
                self.write_colored(Color::Instruction, "branch");
                self.buffer.push_str("     ");
                self.write_colored(Color::Reference, &self.format_instr_ref(*cond));
                self.buffer.push_str(", ");
                self.write_colored(Color::Reference, &self.format_block_ref(*true_block));
                self.buffer.push_str(", ");
                self.write_colored(Color::Reference, &self.format_block_ref(*false_block));
            }
            Instr::Ret(_, value) => {
                self.write_colored(Color::Instruction, "ret");
                self.buffer.push_str("        ");
                self.write_colored(Color::Reference, &self.format_instr_ref(*value));
            }
            Instr::Upsilon(_, phi_ref, value) => {
                self.write_colored(Color::Instruction, "upsilon");
                self.buffer.push_str("    ");
                self.write_colored(Color::Reference, &self.format_phi_ref(*phi_ref));
                self.buffer.push_str(", ");
                self.write_colored(Color::Reference, &self.format_instr_ref(*value));
            }
            Instr::Phi(_, phi_ref) => {
                self.write_colored(Color::Instruction, "phi");
                self.buffer.push_str("        ");
                self.write_colored(Color::Reference, &self.format_phi_ref(*phi_ref));
            }
            Instr::ConstBool(_, value) => {
                self.write_colored(Color::Instruction, "const_bool");
                self.buffer.push_str(" ");
                self.write_colored(Color::Constant, &value.to_string());
            }
            Instr::ConstI32(_, value) => {
                self.write_colored(Color::Instruction, "const_i32");
                self.buffer.push_str("  ");
                self.write_colored(Color::Constant, &value.to_string());
            }
            Instr::Arg(_, index) => {
                self.write_colored(Color::Instruction, "arg");
                self.buffer.push_str("        ");
                self.write_colored(Color::Constant, &index.to_string());
            }
            Instr::Add(_, left, right) => {
                self.write_colored(Color::Instruction, "add");
                self.buffer.push_str("        ");
                self.write_colored(Color::Reference, &self.format_instr_ref(*left));
                self.buffer.push_str(", ");
                self.write_colored(Color::Reference, &self.format_instr_ref(*right));
            }
            Instr::Eq(_, left, right) => {
                self.write_colored(Color::Instruction, "eq");
                self.buffer.push_str("         ");
                self.write_colored(Color::Reference, &self.format_instr_ref(*left));
                self.buffer.push_str(", ");
                self.write_colored(Color::Reference, &self.format_instr_ref(*right));
            }
            Instr::Neq(_, left, right) => {
                self.write_colored(Color::Instruction, "neq");
                self.buffer.push_str("        ");
                self.write_colored(Color::Reference, &self.format_instr_ref(*left));
                self.buffer.push_str(", ");
                self.write_colored(Color::Reference, &self.format_instr_ref(*right));
            }
        }
    }

    fn print_type_comment(&mut self, meta: Meta) {
        let type_str = match meta.get_type() {
            Type::Void => "void",
            Type::Bool => "bool", 
            Type::I32 => "i32",
        };
        
        // Add padding to align comments
        let padding = 20_i32.saturating_sub(self.buffer.len() as i32 % 100); // Rough alignment
        for _ in 0..padding.max(2) {
            self.buffer.push(' ');
        }
        
        self.write_colored(Color::Comment, "; ");
        self.write_colored(Color::Type, type_str);
    }

    fn format_const_ref(&self, instr_ref: InstrRef) -> String {
        let index = instr_ref.get().abs() as u32;
        format!("c{}", index)
    }

    fn format_instr_ref(&self, instr_ref: InstrRef) -> String {
        let index = instr_ref.get();
        if index < 0 {
            self.format_const_ref(instr_ref)
        } else {
            format!("i{}", index)
        }
    }

    fn format_block_ref(&self, block_ref: BlockRef) -> String {
        format!("b{}", block_ref.get().abs())
    }

    fn format_phi_ref(&self, phi_ref: PhiRef) -> String {
        format!("p{}", phi_ref.get().abs())
    }

    fn write_colored(&mut self, color: Color, text: &str) {
        if self.use_colors {
            self.buffer.push_str(color.as_ansi());
            self.buffer.push_str(text);
            self.buffer.push_str(Color::Reset.as_ansi());
        } else {
            self.buffer.push_str(text);
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::ir::*;

    #[test]
    fn test_empty_code() {
        let code = Code::new();
        let printer = IrPrinter::new(&code);
        let output = printer.print();
        
        insta::assert_snapshot!(output, @"");
    }

    #[test]
    fn test_constants_only() {
        let mut code = Code::new();
        code.push_unpinned(Instr::const_i32(42));
        code.push_unpinned(Instr::const_bool(true));
        
        let printer = IrPrinter::new(&code);
        let output = printer.print();
        
        insta::assert_snapshot!(output, @r###"
        ; === CONSTANTS & UNSCHEDULED ===
        c2:    const_bool true  ; bool
        c1:    const_i32  42  ; i32

        "###);
    }

    #[test]
    fn test_scheduled_instructions() {
        let mut code = Code::new();
        let const_ref = code.push_unpinned(Instr::const_i32(42));
        code.push_pinned(Instr::Label(Meta::new(Type::Void), BlockRef::new(1).unwrap()));
        code.push_pinned(Instr::Print(Meta::new(Type::Void), const_ref));
        code.push_pinned(Instr::Ret(Meta::new(Type::Void), const_ref));
        
        let printer = IrPrinter::new(&code);
        let output = printer.print();
        
        insta::assert_snapshot!(output, @r###"
        ; === CONSTANTS & UNSCHEDULED ===
        c1:    const_i32  42  ; i32

        ; === FUNCTION CODE ===
        ; Block b1:
               print      c1
               ret        c1
        "###);
    }

    #[test] 
    fn test_non_void_instruction_labeling() {
        let mut code = Code::new();
        let const_ref = code.push_unpinned(Instr::const_i32(42));
        code.push_pinned(Instr::Label(Meta::new(Type::Void), BlockRef::new(1).unwrap()));
        let add_ref = code.push_pinned(Instr::Add(Meta::new(Type::I32), const_ref, const_ref));
        code.push_pinned(Instr::Ret(Meta::new(Type::Void), add_ref));
        
        let printer = IrPrinter::new(&code);
        let output = printer.print();
        
        insta::assert_snapshot!(output, @r###"
        ; === CONSTANTS & UNSCHEDULED ===
        c1:    const_i32  42  ; i32

        ; === FUNCTION CODE ===
        ; Block b1:
        i2:    add        c1, c1  ; i32
               ret        i2
        "###);
    }

    #[test]
    fn test_comprehensive_example() {
        let mut code = Code::new();
        
        // Constants
        let const42 = code.push_unpinned(Instr::const_i32(42));
        let _const_true = code.push_unpinned(Instr::const_bool(true));
        let const10 = code.push_unpinned(Instr::const_i32(10));
        
        // Function with multiple blocks
        code.push_pinned(Instr::Label(Meta::new(Type::Void), BlockRef::new(1).unwrap()));
        let add_result = code.push_pinned(Instr::Add(Meta::new(Type::I32), const42, const10));
        let eq_result = code.push_pinned(Instr::Eq(Meta::new(Type::Bool), add_result, const42));
        code.push_pinned(Instr::Branch(
            Meta::new(Type::Void), 
            eq_result, 
            BlockRef::new(2).unwrap(), 
            BlockRef::new(3).unwrap()
        ));
        
        code.push_pinned(Instr::Label(Meta::new(Type::Void), BlockRef::new(2).unwrap()));
        code.push_pinned(Instr::Print(Meta::new(Type::Void), const42));
        code.push_pinned(Instr::Ret(Meta::new(Type::Void), const42));
        
        code.push_pinned(Instr::Label(Meta::new(Type::Void), BlockRef::new(3).unwrap()));
        code.push_pinned(Instr::Ret(Meta::new(Type::Void), const10));
        
        let printer = IrPrinter::new(&code);
        let output = printer.print();
        
        insta::assert_snapshot!(output, @r###"
        ; === CONSTANTS & UNSCHEDULED ===
        c3:    const_i32  10  ; i32
        c2:    const_bool true  ; bool
        c1:    const_i32  42       ; i32

        ; === FUNCTION CODE ===
        ; Block b1:
        i2:    add        c1, c3  ; i32
        i3:    eq         i2, c1  ; bool
               branch     i3, b2, b3
        ; Block b2:
               print      c1
               ret        c1
        ; Block b3:
               ret        c3
        "###);
    }

    #[test]
    fn test_colored_output() {
        let mut code = Code::new();
        let const_ref = code.push_unpinned(Instr::const_i32(42));
        code.push_pinned(Instr::Label(Meta::new(Type::Void), BlockRef::new(1).unwrap()));
        code.push_pinned(Instr::Ret(Meta::new(Type::Void), const_ref));
        
        let printer = IrPrinter::new_colored(&code);
        let output = printer.print();
        
        // Should contain ANSI color codes
        assert!(output.contains("\x1b["));
        assert!(output.contains("const_i32"));
        assert!(output.contains("ret"));
        
        // Test structure without colors using snapshot
        let uncolored_printer = IrPrinter::new(&code);
        let uncolored_output = uncolored_printer.print();
        
        insta::assert_snapshot!(uncolored_output, @r###"
        ; === CONSTANTS & UNSCHEDULED ===
        c1:    const_i32  42  ; i32

        ; === FUNCTION CODE ===
        ; Block b1:
               ret        c1
        "###);
    }
}