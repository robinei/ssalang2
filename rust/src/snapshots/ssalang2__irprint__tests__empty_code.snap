---
source: src/irprint.rs
expression: output
---

